### PR TITLE
feat: add and document libretranslate

### PR DESCRIPTION
**Pull request by SethFalco**
## Proposed changes

Adds [LibreTranslate](https://libretranslate.com/) as a Machine Translation option.
I pretty much just tried to copy DeepL where I could to implement this, hopefully it's fine. ^-^'

Closes https://github.com/WeblateOrg/weblate/issues/5830

## Checklist

- [x] Lint and unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added documentation to describe my feature.
- [x] I have squashed my commits into logic units.
- [x] I have described the changes in the commit messages.

PR-ing this already, but in a draft state as I'm struggling to run the tests in Docker.
I have run Weblate in general with the changes though, and can vouch LibreTranslate is working.

Discussion here: https://github.com/WeblateOrg/weblate/discussions/6200

### CODE REVIEW

File: weblate/machinery/models.py [hunk 1]
```diff
@@ -70,6 +70,10 @@ class WeblateConf(AppConf):
     DEEPL_KEY = None
     DEEPL_API_URL = "https://api.deepl.com/v2/"
 
+    # API key for DeepL API
```
Comment (nijel) [inline weblate/machinery/models.py:73]:
```suggestion
    # API key for LibreTranslate
```

File: docs/admin/config.rst [hunk 2]
```diff
@@ -1300,6 +1301,38 @@ API key for the DeepL API, you can register at https://www.deepl.com/pro.html
 
    :ref:`deepl`, :ref:`machine-translation-setup`, :ref:`machine-translation`
 
+.. setting:: MT_LIBRETRANSLATE_API_URL
+
+MT_LIBRETRANSLATE_API_URL
+-------------------------
+
+.. versionadded:: 4.8
```
Comment (nijel) [inline docs/admin/config.rst:1309]: I think it can go into 4.7.1:

```suggestion
.. versionadded:: 4.7.1
```

File: docs/admin/machine.rst [hunk 3]
```diff
@@ -114,6 +114,28 @@ Turn on this service by adding ``weblate.machinery.deepl.DeepLTranslation`` to
     `DeepL pricing <https://www.deepl.com/pro>`_,
     `DeepL API documentation <https://www.deepl.com/api.html>`_
 
+.. _libretranslate:
+
+LibreTranslate
+--------------
+
+.. versionadded:: 4.8
```
Comment (SethFalco) [inline docs/admin/machine.rst:122]:
```suggestion
.. versionadded:: 4.7.1
```

File: docs/admin/config.rst [hunk 4]
```diff
@@ -1300,6 +1301,38 @@ API key for the DeepL API, you can register at https://www.deepl.com/pro.html
...
+
+.. seealso::
+
+   :ref:`libretranslate`, :ref:`machine-translation-setup`, :ref:`machine-translation`
+
+.. setting:: MT_LIBRETRANSLATE_KEY
+
+MT_LIBRETRANSLATE_KEY
+---------------------
+
```
Comment (nijel) [inline docs/admin/config.rst:1329]:
```suggestion
.. versionadded:: 4.7.1

```

**Review by nijel — COMMENTED** [review]
Can you please also add a changelog entry to docs/changes.rst?

File: weblate/machinery/tests.py [hunk 5]
```diff
@@ -1023,6 +1042,62 @@ def test_cache(self):
         self.assertEqual(len(responses.calls), 0)
 
 
+@override_settings(MT_LIBRETRANSLATE_API_URL="URL")
+class LibreTranslateTranslationTest(BaseMachineTranslationTest):
+    MACHINE_CLS = LibreTranslateTranslation
+    EXPECTED_LEN = 1
+    ENGLISH = "EN"
+    SUPPORTED = "ES"
+    NOTSUPPORTED = "CS"
```
Comment (nijel) [inline weblate/machinery/tests.py:1051]: These should be language codes as exposed by the service, so probably:

```suggestion
    ENGLISH = "en"
    SUPPORTED = "es"
    NOTSUPPORTED = "cs"
```

File: weblate/machinery/tests.py [hunk 6]
```diff
@@ -1023,6 +1042,62 @@ def test_cache(self):
         self.assertEqual(len(responses.calls), 0)
 
 
+@override_settings(MT_LIBRETRANSLATE_API_URL="URL")
```
Comment (nijel) [inline weblate/machinery/tests.py:1045]: Should be a real URL for tests and should match the one used in `responses.add`:

```suggestion
@override_settings(MT_LIBRETRANSLATE_API_URL="https://libretranslate.com/")
```

**Comment by SethFalco** [comment]
Sorry for the delays here, still having some issues with the tests locally it seems.
I'm using the Docker environment, but it appears even with the `main` branch (without my code), some tests are failing.

https://gist.github.com/SethFalco/cea7c6acc8b76a6a81460a1a3ae6ec8c

I've set it up outside of Docker now, so it'd be easier to debug. To be honest, I'm not 100% sure what the issue is.
Appears to be database related, since the error is regarding an aborted transaction.

**Comment by nijel** [comment]
You can run only related parts of the testsuite (for example `rundev.sh test weblate.machinery`), this is probably just another issue in the rundev helper (hopefully fixed by ca846ea4725f5e8d8fb7e2e0d9bdad26331d32e4).

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**